libr/asm/arch/8051/8051_ass.c

```diff
@@ -958,97 +950,124 @@ static bool mnem_mov_c(char const*const*arg, ut16 pc, ut8**out) {
-		if (!r_str_casecmp (arg[1], "a")) {
-			return singlearg_register (0xf8, arg[0], out);
+
+		if (!r_str_casecmp (arg[1], "@r1") || !r_str_casecmp (arg[1], "[r1]")) {
+			return single_byte_instr (0xe7, out);
 		}
-		if (arg[1][0] == '#') {
-			return singlearg_immediate (
-				0x78 | register_number (arg[0])
-				, arg[1]
-				, out);
+
+		if (arg[1][0] == 'r') {
+			if (R_BETWEEN ('0', arg[1][1], '7')) {
+				return single_byte_instr (0xe8 + arg[1][1]-'0', out);
+			}
 		}
-		return singlearg_direct (0xa8 | register_number (arg[0])
-			, arg[1]
-			, out);
 	}
-	if (!r_str_casecmp (arg[1], "c")) {
-		return singlearg_bit (0x92, arg[0], out);
+
+	if ((!r_str_casecmp (arg[0], "@r0") || !r_str_casecmp (arg[0], "[r0]"))) {
+		if (resolve_immediate (arg[1] + 1, &src_imm)) {
+			return singlearg_immediate (0x76, arg[1], out);
+		} else if (is_direct (arg[1])) {
+			return singlearg_direct (0xa6, arg[1], out);
+		} else if (!r_str_casecmp (arg[1], "a")) {
+			return single_byte_instr (0xf6, out);
+		}
 	}
-	if (!r_str_casecmp (arg[1], "a")) {
-		return singlearg_direct (0xf5,  arg[0], out);
+	if ((!r_str_casecmp (arg[0], "@r1") || !r_str_casecmp (arg[0], "[r1]"))) {
+		if (resolve_immediate (arg[1] + 1, &src_imm)) {
+			return singlearg_immediate (0x77, arg[1], out);
+		} else if (!r_str_casecmp (arg[1], "a")) {
+			return single_byte_instr (0xf7, out);
+		} else if (is_direct (arg[1])) {
+			return singlearg_direct (0xa7, arg[1], out);
+		}
 	}
-	if (is_reg (arg[1])) {
-		return singlearg_direct (0x88 | register_number (arg[1])
-			, arg[0]
-			, out);
+
+	if (arg[0][0] == 'r') {
+		if (R_BETWEEN ('0', arg[0][1], '7')) {
+			if (resolve_immediate (arg[1] + 1, &src_imm)) {
+				return singlearg_immediate (0x78 + arg[0][1]-'0', arg[1], out);
+			} else if (is_direct (arg[1])) {
+				return singlearg_direct (0xa8 + arg[0][1]-'0', arg[1], out);
+			} else if (!r_str_casecmp (arg[1], "a")) {
+				return single_byte_instr (0xf8 + arg[0][1]-'0', out);
+			}
+		}
 	}
-	if (is_indirect_reg (arg[1])) {
-		return singlearg_direct (0x86 | register_number (arg[1])
-			, arg[0]
-			, out);
+
+	if (parse_hexadecimal (arg[0], &dst_imm)) {
+		if (resolve_immediate (arg[1] + 1, &src_imm)) {//TODO: write doublearg_direct_register for 3 bytes instructions
+			//use : "arg[1][0] == '#'" ?
+			(*out)[0] = 0x75;
+			(*out)[1] = dst_imm;
+			(*out)[2] = src_imm & 0x00ff;
+			*out += 3;
+			return true;
+		}
+
+		if (parse_hexadecimal (arg[1], &src_imm)) {
+			(*out)[0] = 0x85;
+			(*out)[1] = src_imm;
+			(*out)[2] = dst_imm;
+			*out += 3;
+			return true;
+		}
+
+		if (!r_str_casecmp (arg[1], "@r0") || !r_str_casecmp (arg[1], "[r0]")) {
+			return singlearg_direct (0x86, arg[0], out);
+		}
+
+		if (!r_str_casecmp (arg[1], "@r1") || !r_str_casecmp (arg[1], "[r1]")) {
+			return singlearg_direct (0x87, arg[0], out);
+		}
+
+		if (arg[1][0] == 'r') {
+			if (R_BETWEEN ('0', arg[1][1], '7')) {
+				return singlearg_direct (0x88 + arg[1][1]-'0', arg[0], out);
+			}
+		}
```

Comment:
these checks seem to duplicate `is_reg()`/`register_number()`. it looks like there ought to be a function that returns 0 for [r0], 1 for [r1], and 2-9 for r0-r7. maybe `reg_offset()`?